### PR TITLE
ACCUMULO-4681 Created Table & Namespace WeakReference map

### PR DESCRIPTION
**Pull request by milleruntime**
Moved the de-dupe ID code out of KeyExtent to Table and Namespace.  Created "of" static methods to access ID objects stored in internal WeakHashMap.  Replaced all occurrences of "new ID" with static "of" methods.  Created 2 new tests for these changes: NamespaceTest and TableTest.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/client/impl/Table.java [hunk 1]
```diff
@@ -16,24 +16,59 @@
...
+        WeakReference<Table.ID> tableIdRef = tableIds.get(tableIdString);
+        if (tableIdRef != null) {
+          tableId = tableIdRef.get();
+          if (tableId != null) {
+            return tableId;
+          }
+        }
+
+        tableId = new ID(tableIdString);
+        tableIds.put(tableIdString, new WeakReference<>(tableId));
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: I was thinking through the logic of this put method (and it's "compute" alternatives), and realized that there's no way to get a strong reference to the key itself from the map. You can only do comparisons. So, there's a chance that when you do a "put" here, it will reuse the old key, but the new value (in the case where the map entry exists, but the WeakReference is empty). The result is that the entry could disappear from the map even though the object in the WeakReference is still being used. In this case, we lose deduplication. There does not appear to be a solution to this (short of writing our own map class, or iterating over the keyset and comparing keys ourselves), and our deduplication is "best effort", so it's not a big deal... but it's worth keeping in mind, so we don't even try to use this paradigm for guaranteed deduplication.

Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: To clarify what I mean, "put" doesn't replace a key if it already exists. It only updates the value. This also might be mitigated by using Table.ID as the key, since it's less likely that the Table.ID will exist as a key in the map and point to an empty WeakReference.

File: core/src/main/java/org/apache/accumulo/core/client/impl/Table.java [hunk 2]
```diff
@@ -16,24 +16,59 @@
...
   /**
    * Object representing an internal table ID. This class was created to help with type safety. For help obtaining the value of a table ID from Zookeeper, see
    * {@link Tables#getTableId(Instance, String)}
+   *
+   * Uses an internal WeakHashMap and private constructor for storing a WeakReference of every Table.ID. Therefore, a Table.ID can't be instantiated outside
+   * this class and is accessed by calling Table.ID.{@link #of(String)}.
    */
   public static class ID extends AbstractId {
     private static final long serialVersionUID = 7399913185860577809L;
+    static final WeakHashMap<String,WeakReference<Table.ID>> tableIds = new WeakHashMap<>();
```
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: I did see this happen while tinkering around with the Tests.  The keys were staying around but the WeakReference value was null.

File: core/src/main/java/org/apache/accumulo/core/client/impl/Table.java [hunk 3]
```diff
@@ -16,24 +16,59 @@
...
+     *
+     * @param canonical
+     *          table ID string
+     * @return Table.ID object
+     */
+    public static Table.ID of(final String canonical) {
+      return dedupeTableId(canonical);
+    }
+
+    private static Table.ID dedupeTableId(String tableIdString) {
```
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: I thought about this but it would require reflection or passing a pre-constructed instance object as a parameter to the method.  I wanted to avoid reflection since we lose some of the type safety. I also wanted to avoid created objects every time for parameters since they may not be needed if the map already contains the id.  

Passing in a constructor object as a parameter may be a nice compromise but this introduces a bunch of potential exceptions:
<pre>
try {
        id = constructor.newInstance(idString);
      } catch (InstantiationException e) {
        e.printStackTrace();
      } catch (IllegalAccessException e) {
        e.printStackTrace();
      } catch (InvocationTargetException e) {
        e.printStackTrace();
      }
      idHashMap.put(idString, new WeakReference<>(id));
</pre>

File: core/src/main/java/org/apache/accumulo/core/client/impl/Table.java (hunk 1 above)
Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: Perhaps a WeakHashMap isn't the correct data structure for our situation.  Aren't we trying to simply cache IDs to prevent unnecessary object creation?  Perhaps a cache is all we want.

Comment (milleruntime) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: These comments made me think this: https://stackoverflow.com/questions/1802809/javas-weakhashmap-and-caching-why-is-it-referencing-the-keys-not-the-values/1803213

File: core/src/main/java/org/apache/accumulo/core/client/impl/Table.java (hunk 3 above)
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: Passing in a `Function<String,T>` might be a nice compromise. Or a `Supplier<T>`:

```java
protected static <T extends AbstractId> T dedupe(WeakHashMap<String,WeakReference<T>> cache, Supplier<T> newInstanceFunction) {
    //....
}
```
In Table.ID:
```java
public ID of(String canonical) {
    return dedupe(cache, () -> new ID(canonical));
}
```

Shared implementation. No reflection. No unnecessary construction.

File: core/src/main/java/org/apache/accumulo/core/client/impl/Table.java (hunk 1 above)
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: Talked to @keith-turner about this briefly. I think the `WeakHashMap<String,WeakReference<>>` is the right type, but the String that's used as a key should be a new one, so its presence in the map is not dependent on the caller's strong reference to the key, but on the caller's strong reference to the returned Table.ID object. In other words, it should be:

```java
    T ret;
    WeakReference<T> ref = cache.get(canonical);
    if (ref == null || (ret = ref.get()) == null) {
        String s = new String(canonical);
        cache.put(s, ret = new Table.ID(s));
    }
    return ret;
```

That way, the new String is only contained inside the ID and its presence in the map is contingent on the ID object still having a strong reference to it. Of course, the user could call `ID.getCanonical()` and get a strong reference to the key, thus holding it in the map even when the ID object is gone. This could be avoided by the constructor creating a second copy... the first is held by the ID only for the purposes of having a strong reference to the map key, and the second would be returned in `.getCanonical()` and `.toString()`.

Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/impl/Table.java]: The main point is that these objects, even the extra String instances, would only be created when needed. Usually, the previous instance would be used. Still can't avoid the potential duplicates I mentioned earlier, unless you force removal of the old value before calling `put` in the case where `ref != null && ref.get() == null`, but the new String instances should dramatically decrease the liklihood of this happening, and a few dupes are probably okay.

**Comment by ctubbsii** [comment]
This whole thing might be easier to implement with [`CacheBuilder.weakValues()`][1] from Guava. Eviction seems to occur if value disappears. I **think** this will do what we want, and it should automatically hide the WeakReference stuff from us, auto-evicting stuff when the value is gone so we never see it as present in the map. The key can be String, without the need to do any crazy `new String` stuff, or early construction. It ensures a dependency on Guava, though.

[1]: https://google.github.io/guava/releases/16.0/api/docs/com/google/common/cache/CacheBuilder.html#weakValues()

File: core/src/test/java/org/apache/accumulo/core/client/impl/TableTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    table1 = Table.ID.of(tableString);
+    assertEquals(4, Table.ID.tableIds.size());
+    assertEquals(tableString, table1.canonicalID());
+  }
+
+  @Test
+  public void testWeakHashMapDecreases() {
+    // get bunch of table IDs that are not used
+    for (int i = 0; i < 1000; i++)
+      Table.ID.of(new String("table" + i));
```
Comment (mjwall) [inline core/src/test/java/org/apache/accumulo/core/client/impl/TableTest.java]: what is this testing?

Comment (mjwall) [inline core/src/test/java/org/apache/accumulo/core/client/impl/TableTest.java]: oh, I see it in the cleanup

Comment (milleruntime) [inline core/src/test/java/org/apache/accumulo/core/client/impl/TableTest.java]: Yeah I found that was the best way to get the GC to collect.  I will add some comments since this isn't clear.

**Comment by milleruntime** [comment]
> This whole thing might be easier to implement with CacheBuilder.weakValues() from Guava. Eviction seems to occur if value disappears. I think this will do what we want, and it should automatically hide the WeakReference stuff from us, auto-evicting stuff when the value is gone so we never see it as present in the map. The key can be String, without the need to do any crazy new String stuff, or early construction. It ensures a dependency on Guava, though.

This is the way I was leaning yesterday after reading that stackoverflow response.  I think the issue with WeakHashMap is that they Keys are weak and not the values. So we have to add another layer of WeakRefernce for the values, complicating things.  In our case, it is just a cache and we really want weak values to store the ID objects.

I also thought of still using WeakHashMap but flipping the keys and values - `WeakHashMap<AbstractId, String>.`  Using ID as the key so it will automatically be weak and then I guess the canonical String as the value.  But then the value is pretty useless.

**Comment by milleruntime** [comment]
0f083e0c1793e6093e1fcd2fc18faaf8686e7af1 seems to be the most elegant solution.  Guava dependency is already in core so shouldn't pull in any new dependencies.  I will test momentarily if others agree.

File: core/src/main/java/org/apache/accumulo/core/data/impl/KeyExtent.java [hunk 5]
```diff
@@ -60,28 +58,11 @@
...
-      return tableId;
-    }
-  }
-
   private Table.ID tableId;
   private Text textEndRow;
   private Text textPrevEndRow;
 
-  private final Table.ID EMPTY_ID = new Table.ID("");
+  private final Table.ID EMPTY_ID = Table.ID.of("");
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/data/impl/KeyExtent.java]: This could be static.

File: core/src/main/java/org/apache/accumulo/core/data/impl/KeyExtent.java [hunk 6]
```diff
@@ -173,9 +154,9 @@ public KeyExtent(Text flattenedExtent, Text prevEndRow) {
   public void setTableId(Table.ID tId) {
 
     if (tId == null)
-      throw new IllegalArgumentException("null table name not allowed");
+      throw new IllegalArgumentException("null table id not allowed");
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/data/impl/KeyExtent.java]: This could be `Objects.requiresNonNull(tId, "null table id not allowed")`

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ */
+public class NamespaceTest {
+
+  @Test
+  public void testWeakHashMapIncreases() {
+    String namespaceString = "namespace-testWeakHashMapIncreases";
+    Integer initialSize = Namespace.ID.cache.asMap().size();
+    Namespace.ID nsId = Namespace.ID.of(namespaceString);
+    assertEquals(initialSize + 1, Namespace.ID.cache.asMap().size());
+    assertEquals(namespaceString, nsId.canonicalID());
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: Could also test that it does not increase after calling again with the same parameter (`Namespace.ID nsId2 = ....`), and that `assertSame(nsId, nsId2)`.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  @Test
+  public void testWeakHashMapNoDuplicates() {
+    String namespaceString = "namespace-testWeakHashMapNoDuplicates";
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: Could guarantee these strings are unique using JUnit `@Rule TestName`.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    Integer initialSize = Namespace.ID.cache.asMap().size();
+    Namespace.ID nsId = Namespace.ID.of(namespaceString);
+    assertEquals(initialSize + 1, Namespace.ID.cache.asMap().size());
+    assertEquals(namespaceString, nsId.canonicalID());
+
+    // ensure duplicates are not created
+    Namespace.ID builtInNamespaceId = Namespace.ID.of("+accumulo");
+    assertEquals(Namespace.ID.ACCUMULO, builtInNamespaceId);
+    builtInNamespaceId = Namespace.ID.of("+default");
+    assertEquals(Namespace.ID.DEFAULT, builtInNamespaceId);
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: These should be `assertSame` instead of `assertEquals`.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    nsId = Namespace.ID.of(namespaceString);
+    assertEquals(initialSize + 1, Namespace.ID.cache.asMap().size());
+    assertEquals(namespaceString, nsId.canonicalID());
+  }
+}
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: Could add a small test to ensure size decreases once strong refs go away. Might not be able to use `.size()` because Cache may not clean up immediately, but should be able to use `.stream().count()` to get an accurate size of remaining elements after GC.

File: server/master/src/main/java/org/apache/accumulo/master/tableOps/Utils.java [hunk 11]
```diff
@@ -59,21 +57,16 @@ static void checkTableDoesNotExist(Instance instance, String tableName, Table.ID
       throw new AcceptableThriftTableOperationException(null, tableName, operation, TableOperationExceptionType.EXISTS, null);
   }
 
-  static <T extends AbstractId> T getNextTableId(String tableName, Instance instance, Class<T> idClassType) throws AcceptableThriftTableOperationException {
+  static String getNextTableId(String tableName, Instance instance) throws AcceptableThriftTableOperationException {
```
Comment (ctubbsii) [inline server/master/src/main/java/org/apache/accumulo/master/tableOps/Utils.java]: If you wanted to keep the original return type, you could pass in a function to map from String to ID:

```java
static <T extends AbstractId> T getNextId(String name, Instance instance, Function<String, T> newIdFunction) throws AcceptableThriftTableOperationException {
    // ...
    return newIdFunction.apply(new String(nid, UTF_8));
}
```

Can use it like this:
```java
Table.ID tableId = Util.getNextId(tableName, instance, Table.ID::of);
```

File: server/master/src/main/java/org/apache/accumulo/master/tableOps/Utils.java [hunk 12]
```diff
@@ -59,21 +57,16 @@ static void checkTableDoesNotExist(Instance instance, String tableName, Table.ID
...
-          nextId = nextId.add(BigInteger.ONE);
-          return nextId.toString(Character.MAX_RADIX).getBytes(UTF_8);
-        }
+      byte[] nid = zoo.mutate(ntp, ZERO_BYTE, ZooUtil.PUBLIC, currentValue -> {
+        BigInteger nextId = new BigInteger(new String(currentValue, UTF_8), Character.MAX_RADIX);
+        nextId = nextId.add(BigInteger.ONE);
+        return nextId.toString(Character.MAX_RADIX).getBytes(UTF_8);
       });
-      Constructor<T> constructor = idClassType.getConstructor(String.class);
-      return constructor.newInstance(new String(nid, UTF_8));
```
Comment (ctubbsii) [inline server/master/src/main/java/org/apache/accumulo/master/tableOps/Utils.java]: Very big +1 to getting rid of this reflection implementation.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java (hunk 9 above)
Comment (milleruntime) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: Cool.  I've never used `assertSame`

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java (hunk 8 above)
Comment (milleruntime) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: Cool never knew about this.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java (hunk 10 above)
Comment (milleruntime) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: I didn't have much luck measuring `.size()` or `asMap.size()` after GC using their cleanup method.  I will try again using streams.

File: server/master/src/main/java/org/apache/accumulo/master/tableOps/Utils.java (hunk 11 above)
Comment (milleruntime) [inline server/master/src/main/java/org/apache/accumulo/master/tableOps/Utils.java]: Other than using the Function interface, this doesn't seem all that different from the reflection code that I just removed.

Comment (ctubbsii) [inline server/master/src/main/java/org/apache/accumulo/master/tableOps/Utils.java]: It's not different, but it prevents passing a String around (which I believe was the original reason you added the reflection code), but without reflection (lambdas are faster). It's up to you. I like the elegance of the functional stuff, but I don't think it matters much, as long as we're avoiding reflection.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java [hunk 13]
```diff
@@ -17,38 +17,73 @@
...
+    System.out.println("Namespace.ID.cache.asMap().entrySet().stream().count() = " + initialSize);
+    do {
+      System.gc();
+      try {
+        Thread.sleep(5000);
+      } catch (InterruptedException e) {
+        fail("Thread interrupted while waiting for GC");
+      }
+      postGCSize = Namespace.ID.cache.asMap().entrySet().stream().count();
+      System.out.println("After GC: Namespace.ID.cache.asMap().entrySet().stream().count() = " + postGCSize);
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: Should use logger or remove these debugging lines.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java [hunk 14]
```diff
@@ -17,38 +17,73 @@
...
+  @Test(timeout = 60_000)
+  public void testCacheDecreasesAfterGC() {
+    generateJunkCacheEntries();
+    Long initialSize = Namespace.ID.cache.asMap().entrySet().stream().count();
+    Long postGCSize;
+    System.out.println("Namespace.ID.cache.asMap().entrySet().stream().count() = " + initialSize);
+    do {
+      System.gc();
+      try {
+        Thread.sleep(5000);
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: Could shorten this to 1 second or 500ms, since there's a good chance it the condition will be met quickly.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java [hunk 15]
```diff
@@ -17,38 +17,73 @@
...
     assertEquals(namespaceString, nsId.canonicalID());
+    Namespace.ID nsId2 = Namespace.ID.of(namespaceString);
+    assertEquals(initialSize + 1, Namespace.ID.cache.asMap().entrySet().stream().count());
+    assertSame(nsId, nsId2);
+  }
+
+  @Test(timeout = 60_000)
+  public void testCacheDecreasesAfterGC() {
+    generateJunkCacheEntries();
+    Long initialSize = Namespace.ID.cache.asMap().entrySet().stream().count();
```
Comment (ctubbsii) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: This could have already garbage collected some of the entries, after leaving the method, so further GC may not do anything. Instead, consider setting the initialSize before the junk entries are added, and the loop condition should be: `while (postGCSize > initialSize);`

Comment (milleruntime) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: Ah yeah good thinking.

File: core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java (hunk 14 above)
Comment (milleruntime) [inline core/src/test/java/org/apache/accumulo/core/client/impl/NamespaceTest.java]: My original GC test had a finite number of loops but that is a good point, no need to wait longer than needed.
